crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/WebCrawler.java

```diff
@@ -335,8 +310,11 @@ public void run() {
                         }
                         if (curURL != null) {
                             curURL = handleUrlBeforeProcess(curURL);
-                            processPage(curURL);
-                            frontier.setProcessed(curURL);
+                            if (curURL != null) {
+                                boolean processedSuccess = processPage(curURL);
```

Comment:
+1
I tend to use final a lot, and it gives me extra readability somehow, but since most developers are not used to this, I sometimes leave it out when working in their code base.